### ISSUE TITLE
UI Components Library

### ISSUE THREAD

**Issue by ahmadabdolsaheb** [freeCodeCamp/freeCodeCamp#44668]
## Description:
We need a better way of managing our styles and components. In order to achieve that we should create our own light-weight component library that uses Storybook (isolate the components and document them) and tailwindCSS (style each component) and make sure it meets the [web accessibility requirements](https://www.w3.org/WAI/standards-guidelines/wcag/).

## Todo list:
<details>
  <summary>Completed tasks</summary>

- - [x] Setup Storybook - https://github.com/freeCodeCamp/freeCodeCamp/pull/41920
- - [x] Setup Tailwind CSS 
-   - [x] Basic Tailwind config - https://github.com/freeCodeCamp/freeCodeCamp/pull/42153
-   - [x] Add color tokens to tailwind config (split to issue https://github.com/freeCodeCamp/freeCodeCamp/issues/42186 for further discussion) - https://github.com/freeCodeCamp/freeCodeCamp/pull/42388
-   - [x] Support dark mode #43616
- - [x] Setup React Testing Library and its ESLint plugins - https://github.com/freeCodeCamp/freeCodeCamp/pull/42221
- - [x] Setup TypeScript - #43541
- - [x] Setup Prettier & other ESLint rules - https://github.com/freeCodeCamp/freeCodeCamp/pull/42818
- - [x] Package build and publication setup 
-   - [x] Setup Bundler (rollup/webpack/parcel) - https://github.com/freeCodeCamp/freeCodeCamp/pull/42549
-   - [ ] Setup Publication (not for now)
- - [x] Write a guide for component development - @huyenltnguyen
-   - [x] Alert #43835
</details>

- [ ] Setup css linter for custom CSS - related https://github.com/freeCodeCamp/freeCodeCamp/issues/16445
- [ ] Write a guide for component development - @huyenltnguyen
- [ ] Build the following components
  - [x] Alert #43835
  - [ ] Button - @huyenltnguyen 
  - [ ] Col
  - [ ] ControlLabel
  - [ ] DropdownButton (maybe consider using [Headless UI's Menu](https://headlessui.dev/react/menu))
  - [ ] Form
  - [ ] FormControl
  - [ ] FormGroup
  - [ ] Grid
  - [ ] HelpBlock
  - [ ] Image #44590
  - [ ] Media
  - [ ] MenuItem
  - [ ] Modal (maybe consider using [Headless UI's Dialog](https://headlessui.dev/react/dialog))
  - [ ] Panel
  - [ ] Row
  - [ ] Tab
  - [ ] TabPane
  - [ ] Tabs
  - [ ] Table
  - [ ] ToggleButton (maybe consider using [Headless UI's Switch](https://headlessui.dev/react/switch))
  - [ ] ToggleButtonGroup
- [ ] Remove the ToggleButton in `/components/helpers` and replace it with the one above (? - I'm not sure, I haven't looked at the code closely, but I guess maybe there is a reason to split them)
- [ ] Use `npm-link` to link the ui folder with /client
- [ ] Replace `react-bootstrap` components with our own, one component at a time
- [ ] Audit and clean up any unused CSS code
- [ ] Remove `@freecodecamp/react-bootstrap` as a dependency
- [ ] Remove the bootstrap stylesheet injection

### Research
* Find component usage in the project and list all the props being used.
* List out all of the `classNames` passed as a prop
* Compare usage and props with React-Bootstrap documentation.
* Find out if a component has a child component and how they work with regards to passed props.
* Find out which Bootstrap classes are applied by default.
* Find which Bootstrap classes are being used in tests and custom styles.
* See how the react Bootstrap 3 has been created.
* See how the current react Bootstrap component has been created.

### Development
* Initialize Alert component and tests by using the current react Bootstrap implemented as a reference.
* Create an interface for the props.
* Add `children` prop
* Add `className` prop
* convert custom and default styles to tailwind classes
* assign classes based on props
* avoid creating a .css file 

### Useful links
- [React Bootstrap v0.33 src/Alert.js](https://github.com/react-bootstrap/react-bootstrap/blob/bs3-dev/src/)
- [React Bootstrap v0.33 Docs](https://react-bootstrap-v3.netlify.app/components/)
- [Bootstrap 3.3.7 styles](https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.css)
- [React Bootstrap current implementation](https://github.com/react-bootstrap/react-bootstrap/tree/master/src)
- [React Bootstrap current tests](https://github.com/react-bootstrap/react-bootstrap/tree/master/test)

Note:  After the above-mentioned components are used released to production, the rest of the UI components will be added to the library and all CSS will be removed from the client directory.

Note: This issue is based on discussions in #41580

**Comment by ahmadabdolsaheb** [comment]
@sitek94, @huyenltnguyen, @Ismailtlem here is the new home for our UI component lib :)

**Comment by hectoraldairah** [comment]
Hello! I'd like to contribute to this project. ✋ 
Where I can find the link for the component development guide? 🤔

**Comment by Ismailtlem** [comment]
Hi @hectoraldairah 

Have you checked the first comment on this issue ? Is this what you're looking for ?

**Comment by ahmadabdolsaheb** [comment]
@hectoraldairah, here is the link to the docs: https://contribute.freecodecamp.org/#/how-to-work-on-the-component-library

**Comment by Jmagero** [comment]
Hello! I'd like to contribute to this project. ✋

**Comment by Aarush-wadhia** [comment]
Hello, I would like to contribute to this project.

**Comment by luka2220** [comment]
Hello! I would like to help contribute to this project 🙂

**Comment by ahmadabdolsaheb** [comment]
@luka2220 please refer to the Component Library documentation link in the issue description above.

**Comment by shootermv** [comment]
Hi! Can i try to work on FdropdownButton component?

**Comment by ahmadabdolsaheb** [comment]
@shootermv, yes please.

**Comment by masterX89** [comment]
Hi! I found these components like `DropdownButton` , `Modal` and `ToggleButton` are considered using Headless UI. So I wonder if `Tabs`, `Tab` and `TabPane` will also consider [Headless UI's Tabs](https://headlessui.com/react/tabs). 
Can I try to work on these above components?

**Comment by ahmadabdolsaheb** [comment]
@masterX89, we consider using headlessUI if the component is complex and using headless UI would speed up the implementation considerably. 
Feel free to research a component that you are interested and summarize your findings in a new issue. Then, we could discuss the implementation in more detail.

**Comment by mriiad** [comment]
Hello @ahmadabdolsaheb, I would like to work on Panel & MenuItem please!

**Comment by ahmadabdolsaheb** [comment]
Thanks for your interest @mriiad.
Please feel free to proceed with reading the issue and related documentation thoroughly and opening an issue with your research findings on the `Panel` component. 

Let's hold of on the `Menuitem` since it is very close to what @shootermv is working on in #47043.

**Comment by lexyd** [comment]
@ahmadabdolsaheb Is there a design file like on Figma where the consistency of the UI components can be designed first.

I would like to contribute to it.

**Comment by ahmadabdolsaheb** [comment]
@lexyd at this point we are focusing on replacing the existing customized react-bootstrap library. 

Once that is done, we will audit our component library's design. We could consider having Figma files then, but it would be more work to keep them in-sync over time.

I will tag you if we decide to go down that route. Meanwhile, feel free to work on the remaining components if you are interested.

**Comment by lexyd** [comment]
@ahmadabdolsaheb Amazing, ill take a look at the other components

**Comment by JimmyK1995** [comment]
Gotta take part in this amazing project!
Will be taking on the dropdown button, I will write up an issue based on my research, and delete this comment afterwards.

**Comment by ahmadabdolsaheb** [comment]
@JimmyK1995, thank you for your interest. The dropdown is being worked at here: https://github.com/freeCodeCamp/freeCodeCamp/pull/47043, feel free to contact @shootermv on the pull request for pairing or helping with writing the issue. but it does not have an issue yet. Feel free to create an issue so we could finalize the spec or

**Comment by shootermv** [comment]
@ahmadabdolsaheb,  actually i did opened issue already- #47356

**Comment by twhite96** [comment]
I'd like to take a couple of these. Just let me know which ones; I am comfortable with most of these.

**Comment by sauravkumar8178** [comment]
Hello sir, can you assign this issue to me. I am very interested to solving this issue.

**Comment by ahmadabdolsaheb** [comment]
@sauravkumar8178 , please go ahead 👍. No assignment is necessary as described in our [contribution guidelines](https://contribute.freecodecamp.org/#/index?id=where-do-i-start-if-i-want-to-work-on-an-issue). The first comprehensive PR will be considered for review.

**Comment by rajgaur98** [comment]
Hi @ahmadabdolsaheb , I was looking into the `ToggleButton` component. 
I have a question, do we need to implement the `ToggleButton` from `components/helpers` or the `ToggleButton` from `@freecodecamp/react-bootstrap` ?

**Comment by ahmadabdolsaheb** [comment]
@rajgaur98, thanks for your interest. For the first iteration, it should replace the @freecodecamp/react-bootstrap and include the style overwrites.

**Comment by adsingh14** [comment]
After sign-in, 'Alert' didn't auto close. 
![image](https://user-images.githubusercontent.com/35252877/201105508-16c16a31-e8ea-48eb-9ba7-f48547509dee.png)

----
Contribution: **Yes**

**Comment by Sboonny** [comment]
`<Tab>` element doesn't exist in codebase, how should we implement it in the UI_Component?

**Comment by ahmadabdolsaheb** [comment]
@Sboonny, there used to be one, but it has been removed recently. I removed it from the spec accordingly.

**Comment by Sboonny** [comment]
I have marked `ToggleButtonGroup` off, because it has been recently deleted.

**Comment by ensigma** [comment]
If no one's started the modal component. I should be able to knock that out. My nickname was modalBoi when I was on a project at Victoria's Secret 😂

**Comment by kondapalli19** [comment]
I would like to work on this issue can you assign it to me?

**Comment by fluxthedev** [comment]
> I would like to work on this issue can you assign it to me?

Go at it, work and life hit so I won't be able to git to it anytime soon.

**Comment by JordanMooree** [comment]
@kondapalli19 please refer to [this comment](https://github.com/freeCodeCamp/freeCodeCamp/issues/44668#issuecomment-1285088557) - no assignment is necessary

**Comment by winniecwng** [comment]
Hi, there's a comment above that we don't need to be assigned to work on an issue but I just want to make sure that's the case; I'd like to work on the "Audit and clean up any unused CSS" but not sure if this issue will start until all styles are changed to tailwind

**Comment by Sboonny** [comment]
Hi, @winniecwng.

Ideally the CSS properties removal should occur with the swapping of the components, because if we need to revert the component PR, we won't need to revert multiple PRs, the PR that adds the component, and the PR that removes the CSS properties, which can lead to broken styles if the order of reverting is incorrect.

if you want to focus purely on cleaning unused CSS, feel free to open PR that clean unused CSS that you find unused, no need to wait, but waiting can make things easier for you.

**Comment by KSSaiTeja** [comment]
@ahmaxed i want to work on replacing `react-bootstrap`. could you please tell me more about this.

**Comment by ahmaxed** [comment]
@KSSaiTeja, thank you for our interest. The component creation phase has been completed. Now @Sboonny is focused on replacing the components and cleaning out the styles from the client. Feel free to start with replacing the Alert component.

**Comment by KSSaiTeja** [comment]
ok sure.. what should i done for the alert component.. could you please tell me?

**Comment by KSSaiTeja** [comment]
@ahmaxed thanks for the reply. could you please tell me what should i do in alert component

**Comment by KSSaiTeja** [comment]
@ahmaxed should i optimsize the alert component? if you don't mind.. as a beginner i didn't understand what should i do now. could you please tell me clearly what should i do now and mentioning the file you're saying.. 

Hope you will give reply soon..!

**Comment by Sboonny** [comment]
@KSSaiTeja, the workflow should be similar to the `Button` component [PR](https://github.com/freeCodeCamp/freeCodeCamp/pull/51443), and it should be way simpler.

- export the function in `tools/ui-components/src/index.ts`.
- build the development environment. so you can easily import the component through TypeScript.
- remove bootstrap import of the `Alert`.
- import the `Alert` from the library.
- swap the `bsStyle` prop with `variant` prop, TypeScript should warn you if you missed a prop.
- check if the classes related to `.alert` and `.alert-success` are needed, if they aren't needed, remove them.

Note:

There are classes in `bootstrap.min.css` but they shouldn't be changed, because the file will be deleted later when the components are merges.

**Comment by KSSaiTeja** [comment]
@Sboonny thanks for the reply. i will follow your instructions and implement them. And will raise a PR soon

**Comment by Eleazar23** [comment]
Hi, I would like to help with this task: Remove @freecodecamp/react-bootstrap as a dependency

**Comment by Sboonny** [comment]
`Image` and `ToggleButton` aren't used in the `/learn`, we can decide what to do with them later, for now I won't create a PR that add them to `/learn`.
